src/metabase/domain_entities/core.clj

```diff
@@ -0,0 +1,105 @@
+(ns metabase.domain-entities.core
+  (:require [clojure.string :as str]
+            [medley.core :as m]
+            [metabase.domain-entities.specs :refer [domain-entity-specs MBQL]]
+            [metabase.mbql.util :as mbql.u]
+            [metabase.models
+             [card :refer [Card]]
+             [table :as table :refer [Table]]]
+            [metabase.util :as u]
+            [schema.core :as s]))
+
+(def ^:private ^{:arglists '([field])} field-type
+  "Return the most specific type of a given field."
+  (some-fn :special_type :base_type))
+
+(def SourceName
+  "A reference to a `SourceEntity`."
+  s/Str)
+
+(def ^:private DimensionReference s/Str)
+
+(def DimensionBindings
+  "Mapping from dimension name to the corresponding instantiated MBQL snippet"
+  {DimensionReference MBQL})
+
+(def SourceEntity
+  "A source for a card. Can be either a table or another card."
+  (s/cond-pre (type Table) (type Card)))
+
+(def Bindings
+  "Top-level lexial context mapping source names to their corresponding entity and constituent dimensions. See also `DimensionBindings`."
+  {SourceName {(s/optional-key :entity)     SourceEntity
+               (s/required-key :dimensions) DimensionBindings}})
+
+(s/defn ^:private get-dimension-binding :- MBQL
+  [bindings :- Bindings, source :- SourceName, dimension-reference :- DimensionReference]
+  (let [[table-or-dimension maybe-dimension] (str/split dimension-reference #"\.")]
+    (if maybe-dimension
+      (cond->> (get-in bindings [table-or-dimension :dimensions maybe-dimension])
+        (not= source table-or-dimension) (vector :joined-field table-or-dimension))
+      (get-in bindings [source :dimensions table-or-dimension]))))
+
+(s/defn resolve-dimension-clauses
+  "Instantiate all dimension reference in given (nested) structure"
+  [bindings :- Bindings, source :- SourceName, obj]
+  (mbql.u/replace obj
+    [:dimension dimension] (->> dimension
+                                (get-dimension-binding bindings source)
+                                (resolve-dimension-clauses bindings source))))
+
+(s/defn mbql-reference :- MBQL
+  "Return MBQL clause for a given field-like object."
+  [{:keys [id name base_type]}]
+  (if id
+    [:field-id id]
+    [:field-literal name base_type]))
+
+(defn- has-attribute?
+  [entity {:keys [field domain_entity has_many]}]
+  (cond
+    field (some (fn [col]
+                  (or (isa? (field-type col) field)
+                      (= (:name col) (name field))))
```

Comment:
This is the part I'm unsure about. Allowing matching by name opens a whole new can of worms but it's useful in 2 scenarios: 
1) transforms
2) vendor-specific DEs where you then wouldn't have to add new types to account for all the particulars of their schema

Reply:
should the name comparison be case-insensitive?

Reply:
I think it's fine as you'd only use this to target known names